lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,294 @@
+
+    def __unicode__(self):
+        """
+        String representation of this model.
+        """
+        return u"VisibleBlocks object - hash:{}, raw json:'{}'".format(self.hashed, self.blocks_json)
+
+    @property
+    def blocks(self):
+        """
+        Returns the blocks_json data stored on this model as a list of
+        BlockRecords in the order they were provided.
+        """
+        return BlockRecordSet.from_json(self.blocks_json)
+
+
+class PersistentSubsectionGradeQuerySet(models.QuerySet):
+    """
+    A custom QuerySet, that handles creating a VisibleBlocks model on creation, and
+    extracts the course id from the provided usage_key.
+    """
+    def create(self, **kwargs):
+        """
+        Instantiates a new model instance after creating a VisibleBlocks instance.
+
+        Arguments:
+            user_id (int)
+            usage_key (serialized UsageKey)
+            course_version (str)
+            subtree_edited_date (datetime)
+            earned_all (float)
+            possible_all (float)
+            earned_graded (float)
+            possible_graded (float)
+            visible_blocks (iterable of BlockRecord)
+        """
+        visible_blocks = kwargs.pop('visible_blocks')
+        if not isinstance(visible_blocks, BlockRecordSet):
+            visible_blocks = BlockRecordSet(visible_blocks)
+        try:
+            visible_blocks_model = VisibleBlocks.objects.create_from_blockrecords(blocks=visible_blocks)
+        except IntegrityError:
+            # If an integrity error occurs, the VisibleBlocks model we want to
+            # create already exists.  Use the hash for our current list of
+            # blocks as the foreign key.
+            visible_blocks_hash = visible_blocks.to_hash()
+        else:
+            visible_blocks_hash = visible_blocks_model.hashed
+
+        grade = self.model(
+            course_id=kwargs['usage_key'].course_key,
+            visible_blocks_id=visible_blocks_hash,
+            **kwargs
+        )
+        grade.full_clean()
+        grade.save()
+        return grade
+
+
+class PersistentSubsectionGrade(TimeStampedModel):
+    """
+    A django model tracking persistent grades at the subsection level.
+    """
+
+    class Meta(object):
+        unique_together = (('user_id', 'course_id', 'usage_key'), )
+
+    # primary key will need to be large for this table
+    id = UnsignedBigIntAutoField(primary_key=True)  # pylint: disable=invalid-name
+
+    # uniquely identify this particular grade object
+    user_id = models.IntegerField(blank=False)
+    course_id = CourseKeyField(blank=False, max_length=255)
+    usage_key = UsageKeyField(blank=False, max_length=255)
+
+    # Information relating to the state of content when grade was calculated
+    subtree_edited_date = models.DateTimeField('last content edit timestamp', blank=False)
+    course_version = models.CharField('guid of latest course version', blank=False, max_length=255)
```

Comment:
This field needs to be able to accept blank values, see https://github.com/edx/edx-platform/pull/13093/files/a5829c53d9b431dc1454da331de331584dae4a76#r74271632